scipy/integrate/tests/test_quadpack.py

```diff
@@ -503,6 +503,18 @@ def f(x, y, z):
             error_tolerance=6e-8
         )
 
+    def test_complex(self):
+        def tfunc(x):
+            return np.exp(1j*x)
+
+        assert_quad(quad(tfunc, 0, np.pi/2, complex_func=True),
+                    1+1j, error_tolerance=1e-6)
+
+        full_res = quad(tfunc, 0, np.pi/2, complex_func=True, full_output=True)
+        assert_quad(full_res[:-1],
+                    1+1j, error_tolerance=1e-6)
+        assert set(("real output", "imag output")) == set(full_res[2].keys())
```

Comment:
The part that checks the result of integration is somewhat redundant (because we know that `full_output` does not change the results of integration at all) and the test of the rest is not very strong (because it only checks for the presence of the keys, not their content). Please consider the following test:
* Call `quad` three times with `full_output=True`: once on the complex function, once on a function with only the real part of the complex function, and once on a function with only the complex part of the function. 
* Check that the result is `1 + 1j`
* Check that the real parts (including the real part of the error estimate) of the call to `quad` with the complex function equal the output of the call to `quad` with the real part of the complex function
* Check that the imaginary parts (including the imaginary part of the error estimate) of the call to `quad` with the complex function equal the output of the call to `quad` with the imaginary part of the complex function 

It could start like:
```python3
        kwargs = {'a': 0, 'b': np.pi/2, 'full_output': True}
        res_c = quad(lambda x: np.exp(1j*x), complex_func=True, **kwargs)
        res_r = quad(np.cos, complex_func=False, **kwargs)
        res_i = quad(np.sin, complex_func=False, **kwargs)
        assert_allclose(res_c[0], 1 + 1j)
        assert_allclose(res_c[1], res_r[1] + 1j*res_i[1])
```
Then I think you'd need to iterate through the elements of the rest to check.

Reply:
There's a bit of a problem in the test case. `res_c[2]['real']` and `res_c[2]['imag']` are both single-element tuples, each containing a dictionary, whereas `res_r[2]` and `res_i[2]` are dictionaries.

Something that's not tested is a case in which there is a `message` or `explain`.
![image](https://user-images.githubusercontent.com/6570539/196023674-cf8f7be3-b272-419b-9624-70d93eef2f27.png)

Reply:
Update: Yeah this test:
```
kwargs = {'a': 0, 'b': np.inf, 'full_output': True, 'weight':'cos', 'wvar':1,}
res_c = integrate.quad(lambda x: np.exp(1j*x), complex_func=True, **kwargs)
res_r = integrate.quad(np.cos, complex_func=False, **kwargs)
res_i = integrate.quad(np.sin, complex_func=False, **kwargs)
```
shows one of the most complicated cases: `len(res_r) == 5` yet `len(res_r) == 3`.

I do not love the behavior of `quad` in this case : /

Reply:
This is an interesting test case.  I get different numerical results between the components of `res_c` and `res_r` and `res_i`.  This seems to be rooted in dealing with the weights.  

If I remove `weight` and `wvar` from the arguments, I can test the numerical results.  I can also test the error messages are consistent between the cases.  Upon removing the weights, `quad` no longer returns the error explanations, e.g. `len(res_r) == 4`.  However, the explanations give only a static key to decode the error-codes in the `ierlst` when `weight in [cos, sin]` and `a==Inf or b==Inf`, see [L443](https://github.com/scipy/scipy/blob/main/scipy/integrate/_quadpack_py.py#L443); it is not specific to the error at hand.  

Although I see different numerical results of the integrals, they are within the estimated error of each other.  

Maybe avoiding the numeric inconsistencies gives us a better test of the behavior with minimal trade-off.

Reply:
Here is a summary of the output I get from a terminal; each line of output shows the integral and the estimated error.  

```
In [1]: import numpy as np

In [2]: from scipy.integrate import quad

In [3]: kwargs = {'a': 0, 'b': np.inf, 'full_output': True, 'weight': 'cos', 'wvar': 1}

In [4]: quad(lambda x: np.exp(1j*x), complex_func=True, **kwargs)[:2]
Out[4]:
((230.90706003885032-7.878139118594558e-16j),
 (292.16811678405566+4.827353117680502e-15j))

In [5]: quad(np.cos, **kwargs)[:2]
Out[5]: (61.26105674500084, 292.16811678405566)

In [6]: quad(np.sin, **kwargs)[:2]
Out[6]: (-6.798420242847239e-16, 4.362202832626668e-15)
```

If I remove the cosine weight, it's a different story.
```
In [8]: kwargs = {'a': 0, 'b': np.inf, 'full_output': True}

In [9]: quad(lambda x: np.exp(1j*x), complex_func=True, **kwargs)[:2]
Out[9]:
((19750109.207199536-4695026.791622601j),
 (40229039.04611343+11164570.125259375j))

In [10]: quad(np.cos, **kwargs)[:2]
Out[10]: (19750109.207199536, 40229039.046113424)

In [11]: quad(np.sin, **kwargs)[:2]
Out[11]: (-4695026.7916226005, 11164570.125259371)
```

And, similarly, manually specifying the cosine weight returns consistent results.
```
In [15]: quad(lambda x: np.exp(1j*x)*np.cos(1*x), complex_func=True, **kwargs)[:2]
Out[15]:
((259.3426880212052-4751085.266060017j),
 (114.19578235701343+9343303.762471959j))

In [16]: quad(lambda x: np.cos(x)*np.cos(1*x), **kwargs)[:2]
Out[16]: (259.3426880212057, 114.1957823570026)

In [17]: quad(lambda x: np.sin(x)*np.cos(1*x), **kwargs)[:2]
Out[17]: (-4751085.266060016, 9343303.762471959)
```

I wonder if this may also be related to gh-12662.

Reply:
Hmm. Would you see if they are consistent using `'weight': 'cos'` but changing to `'b': 1`?  That way the integral is not divergent, so there is less chance of bizarre behavior. Do as many of the tests described in https://github.com/scipy/scipy/pull/17051#pullrequestreview-1161074446 as you can for that case. 

Then, switch to the divergent case, and just make sure that the real part of the complex interval gets "message" and "explain" like the real integral does.